File: ios/Sasquatsh/ViewModels/ChatViewModel.swift
import SwiftUI

@Observable
@MainActor
final class ChatViewModel {
    var messages: [ChatMessage] = []
    var messageText = ""
    var isLoading = false
    var isSending = false
    var error: String?
    var hasMore = true

    private var contextType = ""
    private var contextId = ""
    private var services: ServiceContainer?
    private var pollTimer: Timer?

    func configure(services: ServiceContainer, contextType: String, contextId: String) {
        self.services = services
        self.contextType = contextType
        self.contextId = contextId
    }

    func loadMessages() async {
        guard let services, !contextId.isEmpty else { return }
        isLoading = true
        error = nil
        do {
            let fetched = try await services.chat.getMessages(
                contextType: contextType,
                contextId: contextId,
                limit: 50,
                before: nil
            )
            messages = fetched.reversed()
            hasMore = fetched.count >= 50
        } catch {
            self.error = error.localizedDescription
        }
        isLoading = false
    }

    func loadEarlierMessages() async {
        guard let services, let oldest = messages.first else { return }
        do {
            let fetched = try await services.chat.getMessages(
                contextType: contextType,
                contextId: contextId,
                limit: 50,
                before: oldest.createdAt
            )
            let earlier = fetched.reversed()
            messages.insert(contentsOf: earlier, at: 0)
            hasMore = fetched.count >= 50
        } catch {
            self.error = error.localizedDescription
        }
    }

    func sendMessage() async {
        guard let services else { return }
        let content = messageText.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !content.isEmpty, content.count <= 1000 else { return }
        isSending = true
        error = nil
        let text = messageText
        messageText = ""
        do {
            let message = try await services.chat.sendMessage(
                contextType: contextType,
                contextId: contextId,
                content: content
            )
            if !messages.contains(where: { $0.id == message.id }) {
                messages.append(message)
            }
        } catch {
            messageText = text
            self.error = error.localizedDescription
        }
        isSending = false
    }

    func deleteMessage(_ id: String) async {
        guard let services else { return }
        do {
            try await services.chat.deleteMessage(
                contextType: contextType,
                contextId: contextId,
                messageId: id
            )
            messages.removeAll { $0.id == id }
        } catch {
            self.error = error.localizedDescription
        }
    }

    func reportMessage(_ id: String, reason: String, details: String?) async {
        guard let services else { return }
        do {
            try await services.chat.reportMessage(
                contextType: contextType,
                contextId: contextId,
                messageId: id,
                reason: reason,
                details: details
            )
        } catch {
            self.error = error.localizedDescription
        }
    }

    func startPolling() {
        stopPolling()
        pollTimer = Timer.scheduledTimer(withTimeInterval: 10, repeats: true) { [weak self] _ in
            guard let self else { return }
            Task { @MainActor in
                await self.pollNewMessages()
            }
        }
    }

    func stopPolling() {
        pollTimer?.invalidate()
        pollTimer = nil
    }

    private func pollNewMessages() async {
        guard let services, !contextId.isEmpty else { return }
        do {
            let fetched = try await services.chat.getMessages(
                contextType: contextType,
                contextId: contextId,
                limit: 50,
                before: nil
            )
            let reversed = fetched.reversed()
            let existingIds = Set(messages.map(\.id))
            let newMessages = reversed.filter { !existingIds.contains($0.id) }
            if !newMessages.isEmpty {
                messages.append(contentsOf: newMessages)
            }
        } catch {}
    }
}

File: android/app/src/main/java/com/sasquatsh/app/viewmodels/ChatViewModel.kt
package com.sasquatsh.app.viewmodels

import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.sasquatsh.app.models.ChatMessage
import com.sasquatsh.app.services.ChatService
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import javax.inject.Inject

data class ChatUiState(
    val messages: List<ChatMessage> = emptyList(),
    val messageText: String = "",
    val isLoading: Boolean = false,
    val isSending: Boolean = false,
    val error: String? = null,
    val hasMore: Boolean = true
)

@HiltViewModel
class ChatViewModel @Inject constructor(
    private val chatService: ChatService,
    savedStateHandle: SavedStateHandle
) : ViewModel() {

    private val _uiState = MutableStateFlow(ChatUiState())
    val uiState: StateFlow<ChatUiState> = _uiState.asStateFlow()

    private var contextType: String = savedStateHandle["contextType"] ?: ""
    private var contextId: String = savedStateHandle["contextId"] ?: ""
    private var pollJob: Job? = null

    fun configure(contextType: String, contextId: String) {
        this.contextType = contextType
        this.contextId = contextId
    }

    fun updateMessageText(text: String) {
        _uiState.update { it.copy(messageText = text) }
    }

    fun loadMessages() {
        if (contextId.isEmpty()) return
        viewModelScope.launch {
            _uiState.update { it.copy(isLoading = true, error = null) }
            try {
                val fetched = chatService.getMessages(
                    contextType = contextType,
                    contextId = contextId,
                    limit = 50,
                    before = null
                )
                _uiState.update {
                    it.copy(
                        messages = fetched.reversed(),
                        hasMore = fetched.size >= 50,
                        isLoading = false
                    )
                }
            } catch (e: Exception) {
                _uiState.update { it.copy(error = e.localizedMessage, isLoading = false) }
            }
        }
    }

    fun loadEarlierMessages() {
        val oldest = _uiState.value.messages.firstOrNull() ?: return
        viewModelScope.launch {
            try {
                val fetched = chatService.getMessages(
                    contextType = contextType,
                    contextId = contextId,
                    limit = 50,
                    before = oldest.createdAt
                )
                val earlier = fetched.reversed()
                _uiState.update { state ->
                    state.copy(
                        messages = earlier + state.messages,
                        hasMore = fetched.size >= 50
                    )
                }
            } catch (e: Exception) {
                _uiState.update { it.copy(error = e.localizedMessage) }
            }
        }
    }

    fun sendMessage() {
        val content = _uiState.value.messageText.trim()
        if (content.isEmpty() || content.length > 1000) return

        viewModelScope.launch {
            _uiState.update { it.copy(isSending = true, error = null) }
            val originalText = _uiState.value.messageText
            _uiState.update { it.copy(messageText = "") }

            try {
                val message = chatService.sendMessage(
                    contextType = contextType,
                    contextId = contextId,
                    content = content
                )
                _uiState.update { state ->
                    val existingIds = state.messages.map { it.id }.toSet()
                    if (message.id !in existingIds) {
                        state.copy(
                            messages = state.messages + message,
                            isSending = false
                        )
                    } else {
                        state.copy(isSending = false)
                    }
                }
            } catch (e: Exception) {
                _uiState.update {
                    it.copy(
                        messageText = originalText,
                        error = e.localizedMessage,
                        isSending = false
                    )
                }
            }
        }
    }

    fun deleteMessage(id: String) {
        viewModelScope.launch {
            try {
                chatService.deleteMessage(
                    contextType = contextType,
                    contextId = contextId,
                    messageId = id
                )
                _uiState.update { state ->
                    state.copy(messages = state.messages.filter { it.id != id })
                }
            } catch (e: Exception) {
                _uiState.update { it.copy(error = e.localizedMessage) }
            }
        }
    }

    fun reportMessage(id: String, reason: String, details: String?) {
        viewModelScope.launch {
            try {
                chatService.reportMessage(
                    contextType = contextType,
                    contextId = contextId,
                    messageId = id,
                    reason = reason,
                    details = details
                )
            } catch (e: Exception) {
                _uiState.update { it.copy(error = e.localizedMessage) }
            }
        }
    }

    fun startPolling() {
        stopPolling()
        pollJob = viewModelScope.launch {
            while (isActive) {
                delay(10_000) // 10 seconds
                pollNewMessages()
            }
        }
    }

    fun stopPolling() {
        pollJob?.cancel()
        pollJob = null
    }

    private suspend fun pollNewMessages() {
        if (contextId.isEmpty()) return
        try {
            val fetched = chatService.getMessages(
                contextType = contextType,
                contextId = contextId,
                limit = 50,
                before = null
            )
            val reversed = fetched.reversed()
            _uiState.update { state ->
                val existingIds = state.messages.map { it.id }.toSet()
                val newMessages = reversed.filter { it.id !in existingIds }
                if (newMessages.isNotEmpty()) {
                    state.copy(messages = state.messages + newMessages)
                } else state
            }
        } catch (_: Exception) {
            // Polling errors are non-critical
        }
    }

    override fun onCleared() {
        super.onCleared()
        stopPolling()
    }
}
